discovery/rendezvous.go

```diff
@@ -117,6 +117,17 @@ func (r *Rendezvous) Discover(
 			ticker = time.NewTicker(newPeriod)
 		case <-ticker.C:
 			srv := r.servers[rand.Intn(len(r.servers))]
+			// Having a deadline longer than `ticker` period may lead to a situation
+			// when other cases are never checked. Hence, make sure that `period`
+			// is still open.
```

Comment:
but this is why multiple reads in select are randomized. so even if ticker.C is always full - read from period should happen

Reply:
Eventually it will but it can take non-trivial amount of time. I experienced that when testing `eth.staging` fleet. The IP and port can be resolved but the target rendezvous node does not respond so it times out after 5 seconds. If it happens that `case <-ticker.C` is selected 2 times in a row, it's at least 10 seconds before the node shutdowns. 

This solution is still not perfect. What should happen is that a parent context should be canceled even before the timeout and this way the loop would be exited immediately.

Reply:
Yes, I will try to do that.